src/policy/mallocspace/metadata.rs

```diff
@@ -1,61 +1,137 @@
+/// This is a global side metadata spec even though it is used only by MallocSpace as
+/// we require its space to be contiguous and mapped only once. Otherwise we risk
+/// overwriting the previous mapping.
+pub(crate) const ACTIVE_CHUNK_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    is_global: true,
+    offset: GLOBAL_SIDE_METADATA_BASE_ADDRESS.as_usize(),
+    log_num_of_bits: 3,
+    log_min_obj_size: LOG_BYTES_IN_CHUNK as usize,
+};
+
 /// This is the metadata spec for the alloc-bit.
 ///
-/// An alloc-bit is required per min-object-size aligned address , rather than per object, and can only exist as side metadata.
+/// An alloc-bit is required per min-object-size aligned address, rather than per object, and can only exist as side metadata.
 ///
 /// The other metadata used by MallocSpace is mark-bit, which is per-object and can be kept in object header if the VM allows it.
 /// Thus, mark-bit is vm-dependant and is part of each VM's ObjectModel.
 ///
-#[cfg(target_pointer_width = "32")]
 pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: 0,
+    offset: if cfg!(target_pointer_width = "64") {
+        LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize()
+    } else {
+        0
+    },
     log_num_of_bits: 0,
     log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
 };
 
+/// Metadata spec for the active page byte
+///
+/// We use a byte instead of a bit to avoid synchronization costs, i.e. to avoid
+/// the case where two threads try to update different bits in the same byte at
+/// the same time
 #[cfg(target_pointer_width = "64")]
-pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize(),
-    log_num_of_bits: 0,
-    log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_address_range_size(&ALLOC_SIDE_METADATA_SPEC),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
+};
+
+#[cfg(target_pointer_width = "32")]
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    is_global: false,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_bytes_per_chunk(
+            ALLOC_SIDE_METADATA_SPEC.log_min_obj_size,
+            ALLOC_SIDE_METADATA_SPEC.log_num_of_bits,
+        ),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
 };
 
 pub fn is_meta_space_mapped(address: Address) -> bool {
     let chunk_start = conversions::chunk_align_down(address);
-    ACTIVE_CHUNKS.read().unwrap().contains(&chunk_start)
+    is_chunk_mapped(chunk_start) && is_chunk_marked(chunk_start)
+}
+
+fn map_chunk_mark_space(chunk_start: Address) {
+    // We eagerly map 16Gb worth of space for the chunk mark bytes on 64-bits
```

Comment:
Is this still necessary? It seems for every malloc result, you will check if it is mapped. If not, you will invoke this method. And when you create sweep packets, you also check if chunks are mapped, and only mapped chunks will be swept. If that's the case, it seems you do not need the eager mapping here.

Reply:
But it is fine to do this even if it is not necesary. It only allocates 4KB here I guess (16GB/4MB).

Reply:
What is the reason that you need to map N chunks both before and after the address?

Reply:
> What is the reason that you need to map N chunks both before and after the address?

Because `malloc()` is going to allocate memory _around_ the first chunk allocated. So before as well as after the first allocation.

> Is this still necessary? It seems for every malloc result, you will check if it is mapped. If not, you will invoke this method. And when you create sweep packets, you also check if chunks are mapped, and only mapped chunks will be swept. If that's the case, it seems you do not need the eager mapping here.

Well I guess the difference now is that one mutator can do it for likely the rest of the program, but we have to map the metadata for every chunk otherwise? It might introduce unnecessary slowdowns because of the synchronization via Mutex.

Reply:
> CHUNK_METADATA.try_map_metadata_space(): This is intended to use after try_map_metadata_address_range(). Directly using this would overwrite existing mmapping.

Sorry I don't follow? My understanding was that adding the active chunk metadata to the MallocSpace metadata won't overwrite the mapping when `metadata.try_map_metadata_space()` is called.

Reply:
OK. If malloc returns addresses that are not close to the previous ones, this also works, right? It just goes to this method a few more times and map a few more memory.

Reply:
Yeah this works as per the `hsqldb` test for the openjdk binding -- it eagerly allocates the active chunk metadata twice (as seen from the debug messages)

Reply:
> > CHUNK_METADATA.try_map_metadata_space(): This is intended to use after try_map_metadata_address_range(). Directly using this would overwrite existing mmapping.
> 
> Sorry I don't follow? My understanding was that adding the active chunk metadata to the MallocSpace metadata won't overwrite the mapping when `metadata.try_map_metadata_space()` is called as it is already eagerly mapped before.

I was talking about overwriting mmappings with `mmap`. I checked the code, it should be fine. Now when we map an unmapped chunk, we always use mmap with no_replace which will not overwrite existing mappings. Before the change in https://github.com/mmtk/mmtk-core/pull/334, `try_map_metadata_space()` will do a `mmap` and that may overwrite/replace existing mappings. Ignore me on this. Sorry.